pylearn2/sandbox/nlp/models/mlp.py

```diff
@@ -189,3 +198,303 @@ def get_l1_weight_decay(self, coeff):
         assert isinstance(coeff, float) or hasattr(coeff, 'dtype')
         W, = self.transformer.get_params()
         return coeff * abs(W).sum()
+
+
+class ContrastiveProbabilityLayer(Layer):
+    """
+    Predicts the probability of a sample to come from a positive distribution
+    or a negative distribution. For each sample (i.e. a row in the input),
+    this layer produces *two* sets of probabilities. The first set has n
```

Comment:
No, if an example is (x, c) and it can come from either positive distribution D or negative distribution N then we are modeling P(D | x,c) and P(N | x,c).